src/main/java/com/microsoft/sqlserver/jdbc/osgi/Activator.java

```diff
@@ -35,6 +35,9 @@ public void start(BundleContext context) throws Exception {
 
     @Override
     public void stop(BundleContext context) throws Exception {
-        service.unregister();
+        if (service != null) {
+            service.unregister();
+        }
+        SQLServerDriver.deregister();
```

Comment:
This should not be called here since it interfere with the rest of the system because of the static nature. Using statics inside OSGi is often error prone and unnecessary. See comment on SQLServerDriver how this can be fixed.

I also wonder if it would make sence to create exactly one instance of SQLServerDriver (is it thread safe?) and return this one from the SQLServerDataSourceFactory as well.

Reply:
With the previous implementation, the driver was registering itself with DriverManager on `start()`, but never deregistering. This is an issue, because OSGI framework users should not be forced to deregister the JDBC driver manually from their OSGI application. Since the driver is registered on `start()`, it also needs to be deregistered on `stop()` to prevent memory leak. We understand this would affect the entire JVM process, as it did for our JUnit tests, but we believe the correct approach here is to let the applications handle register/deregister calls separately.

Reply:
The problem is, that this only cover the case when one use the bundle and start/stop it, but you can load classes from bundles even in the resolved state. From an OSGi perspective it would be better if the driver won't register at all with the drivermanager.

Another problem arises when you install two different versions, then again, two different driver versions will be register with the manager, so in general there is no clean way in OSGi to handle such static initilizers.

Reply:
Maybe it would be possible to have an abstract class AbstractSQLServerDriver that contains everything except the static initializer-code, the SQLServerDriver then simply extends AbstractSQLServerDriver and adds this static initializer. Then the OSGi Code can have an OSGiSQLServerDriver extending AbstractSQLServerDriver that simply adds nothing at the moment.
That way it would behave perfectly in both worlds, you can use the OSGi-way and you can use the "plain-old way" (even inside an OSGi env) without interfering with each other.

Reply:
By the way I think
```
SQLServerDriver driver = new SQLServerDriver();
properties.put(DataSourceFactory.OSGI_JDBC_DRIVER_CLASS, driver.getClass().getName());
properties.put(DataSourceFactory.OSGI_JDBC_DRIVER_NAME, "Microsoft JDBC Driver for SQL Server");
properties.put(DataSourceFactory.OSGI_JDBC_DRIVER_VERSION,
driver.getMajorVersion() + "." + driver.getMinorVersion());
```
can be replaced by
```
properties.put(DataSourceFactory.OSGI_JDBC_DRIVER_CLASS, SQLServerDriver.class.getName());
properties.put(DataSourceFactory.OSGI_JDBC_DRIVER_NAME, "Microsoft JDBC Driver for SQL Server");
properties.put(DataSourceFactory.OSGI_JDBC_DRIVER_VERSION,
SQLJdbcVersion.major + "." + SQLJdbcVersion.minor);
```
This would avoid creating an unnecessary SQLServerDriver instance just to get the version info

Reply:
@laeubi 

It's not just about getting version info from the driver, the driver needs to be loaded and registered with DriverManager and BundleContext both when Activator > start() is called. This is done in order to make sure that everything running on the current JVM process or WebApp always point to same driver version, it could be JDBC or OSGI API calls, we as library provider always need to make sure that driver is loaded and available for all available services to use.

When the BundleActivator is stopped, it is fair to unload all resources including driver registration from DriverManager as this method is specifically designed for that purpose. You must be aware that BundleActivator is never called concurrently and if another instance of BundleActivator is needed, you must call start() again that will load the driver for you. For any other piece of code using DriverManager or JDBC APIs, must always take into consideration stopping JDBC Driver's Activator and deregistration of driver with it.

> Another problem arises when you install two different versions, then again, two different driver versions will be register with the manager, so in general there is no clean way in OSGi to handle such static initilizers.

I'm sorry but this is just not recommended and neither is a valid scenario. There must never be 2 versions of the same library/driver loaded in one JVM process specified in ClassPath.